File: project-euler/euler58.cpp
#include <iostream>
#include <vector>

#include "utils.hpp"

const std::vector<uint> synteic_spiral(uint n) {
    uint layer = 1;
    uint circle_size = 1;
    uint curr_circle_size = 0;
    uint edge_size = 0;
    std::vector<uint> diagonal_elems;
    for(uint i = 1; i < n; ++i) {
        if(curr_circle_size == circle_size-3*edge_size-1 || 
           curr_circle_size == circle_size-2*edge_size-1 ||
           curr_circle_size == circle_size-edge_size-1 ||
           curr_circle_size == circle_size-1 )
            diagonal_elems.push_back(i);
        if(curr_circle_size == circle_size) {
            curr_circle_size = 0;
            layer += 1;
            edge_size = 2*(layer - 1);
            circle_size = 4*edge_size;
        }
        curr_circle_size += 1;
    }
    return diagonal_elems;
}

int main() {
    uint limit = 26241;
    auto diag = synteic_spiral(limit*limit);
    auto prime_table = fill_prime_table(limit+2);

    uint n_primes = 0;
    for(auto i : diag) {
        bool tmp = is_prime(prime_table, i);
        if( tmp ) {
            n_primes++;
        }
    }
    n_primes--;

    std::cout << n_primes << " " << (2*limit-1) << std::endl;
    std::cout << static_cast<double>(n_primes)/static_cast<double>(2*limit-1) << std::endl;
    return 0;
}

File: project-euler/euler58.py
# -*- coding: utf-8 -*-

# -*- coding: utf-8 -*-

import operator
import numpy as np
from factorize import fill_primes
from utils import index, isprime

limit = 10**7

#primes = fill_primes(10000000)

def syntetic_spiral(n):
    layer = 1
    circle_size = 1
    curr_circle_size = 0
    edge_size = 0
    diag = []
    for i in range(1,n):
        if i % 100000 == 0:
            print(i)
        edges = [circle_size-3*edge_size-1, circle_size-2*edge_size-1, circle_size-edge_size-1, circle_size-1]
        if curr_circle_size in edges:
            diag.append(i)
        
        if curr_circle_size == circle_size:
            curr_circle_size = 0
            layer += 1
            edge_size = 2*(layer - 1)
            circle_size = 4*edge_size
        
        curr_circle_size += 1  
    return diag

def prime_frac(d):
    n_primes = 0
    for i in range(len(d)):
        if i % 100 == 0:
            print(i)
        if isprime(d[i]):
            n_primes += 1
    print(n_primes, len(d))
    return n_primes / len(d)

n = 11
f = prime_frac(syntetic_spiral(n**2))



